### DISCUSSION TITLE
Data source name not found and no default driver specified

### DISCUSSION THREAD

**Discussion by BlaneG** [sqlalchemy/sqlalchemy#8096] [Usage Questions]
Documentation on how to specify the driver for MSSQL hostname-connections would benefit users. 

I was trying to connect to a Microsoft SQL server database using SQLAlchemy.create_engine() (SQLAlchemy=1.3.1), and Python 3.7 on a windows machine with the following connection string:  ```SQLALCHEMY_DATABASE_URI = f'mssql+pyodbc://{SQL_USER_NAME}@{SQL_SERVER}:{SQL_PASSWORD}@{SQL_SERVER}:1433/{SQL_DATABASE}?driver=ODBC+Driver+17+for+SQL+Server'``` as shown in the example [here](https://docs.sqlalchemy.org/en/14/dialects/mssql.html#hostname-connections)  

I noticed the documentation version for SQLAlchemy I was using was different that my installed version, so I also tried this example: ```SQLALCHEMY_DATABASE_URI = f'mssql+pyodbc://{SQL_USER_NAME}@{SQL_SERVER}:{SQL_PASSWORD}@{SQL_SERVER}:1433/{SQL_DATABASE}?driver=SQL+Server+Native+Client+10.0'``` from [here ](https://docs.sqlalchemy.org/en/14/dialects/mssql.html#hostname-connections) which resulted in the same error. 

[This ](https://stackoverflow.com/questions/41344054/python-sqlalchemy-data-source-name-not-found-and-no-default-driver-specified) stackoverflow thread also suggests using ```?driver=SQL+Server` as the driver statement which worked for me.  

Adding links to help determine the correct ODBC driver for mssql would be useful ([for example](https://docs.microsoft.com/en-us/sql/database-engine/configure-windows/check-the-odbc-sql-server-driver-version-windows?view=sql-server-ver16)), and also a resource to specify the correct driver string: 
 https://github.com/mkleehammer/pyodbc/wiki/Connecting-to-SQL-Server-from-Windows

**Comment by zzzeek** [comment]
> Documentation on how to specify the driver for MSSQL hostname-connections would benefit users.

that's the section that you are referring towards in your comment below, which is here: https://docs.sqlalchemy.org/en/14/dialects/mssql.html#hostname-connections

> 
> I was trying to connect to a Microsoft SQL server database using SQLAlchemy.create_engine() (SQLAlchemy=1.3.1), and Python 3.7 on a windows machine with the following connection string: `SQLALCHEMY_DATABASE_URI = f'mssql+pyodbc://{SQL_USER_NAME}@{SQL_SERVER}:{SQL_PASSWORD}@{SQL_SERVER}:1433/{SQL_DATABASE}?driver=ODBC+Driver+17+for+SQL+Server'` as shown in the example [here](https://docs.sqlalchemy.org/en/14/dialects/mssql.html#hostname-connections)
> 
> I noticed the documentation version for SQLAlchemy I was using was different that my installed version, so I also tried this example: `SQLALCHEMY_DATABASE_URI = f'mssql+pyodbc://{SQL_USER_NAME}@{SQL_SERVER}:{SQL_PASSWORD}@{SQL_SERVER}:1433/{SQL_DATABASE}?driver=SQL+Server+Native+Client+10.0'` from [here ](https://docs.sqlalchemy.org/en/14/dialects/mssql.html#hostname-connections) which resulted in the same error.
> 
> [This ](https://stackoverflow.com/questions/41344054/python-sqlalchemy-data-source-name-not-found-and-no-default-driver-specified) stackoverflow thread also suggests using ```?driver=SQL+Server` as the driver statement which worked for me.

I'm not seeing how the query we have in the linked docs is fundamentally different, SO answer:

```py
f"mssql+pyodbc://{uname}:{pword}@{server}:{port}/{dbname}?driver=ODBC+Driver+17+for+SQL+Server"
```
our [docs](https://docs.sqlalchemy.org/en/14/dialects/mssql.html#hostname-connections):

```py
engine = create_engine("mssql+pyodbc://scott:tiger@myhost:port/databasename?driver=ODBC+Driver+17+for+SQL+Server")
```


> 
> Adding links to help determine the correct ODBC driver for mssql would be useful ([for example](https://docs.microsoft.com/en-us/sql/database-engine/configure-windows/check-the-odbc-sql-server-driver-version-windows?view=sql-server-ver16)), and also a resource to specify the correct driver string: https://github.com/mkleehammer/pyodbc/wiki/Connecting-to-SQL-Server-from-Windows

we have links to pyodbc's wiki at https://code.google.com/archive/p/pyodbc/wikis/ConnectionStrings.wiki, if that link is obsolete we can update it

**Reply by BlaneG** [reply]
Thanks for your quick response @zzzeek.  Above I added the details of my issue and how I solved it to provide some context. The solution was that I needed to label the driver as ```driver=SQL+Server``` based on the ODBC driver installed on my system.  It wasn't clear to me how I needed to modify ```f"mssql+pyodbc://{uname}:{pword}@{server}:{port}/{dbname}?driver=ODBC+Driver+17+for+SQL+Server"```.  

Adding information from the last two links that I cited above to the documentation would have helped me to identify what ODBC driver I had installed, and what driver string I needed to specify. Thanks.
